### PR TITLE
Adding back the missing PHPStan implementation

### PR DESCRIPTION
**Pull request by zaherg**
As mentioned here https://github.com/rhubarbgroup/redis-cache/issues/369 when I implemented the new enhanced implementation for testing, I removed the code related to PHPStan.

This PR should add PHPStan back while also provide the ability to auto run it using Github actions.

cc @szepeviktor your feedback is appreciate it before I make it as ready for review.

### CODE REVIEW

**Comment by zaherg** [comment]
@szepeviktor I just wonder, if you don't mind answering, why did you added the `sed` and changed the class name, is there a specific reason for that?

**Comment by szepeviktor** [comment]
Please fix includes/class-plugin.php:207 first. That is a critical error.

**Comment by zaherg** [comment]
got it, thanks for the explanation.

currently, phpstan reports 77 issues at level 5. so it looks like there are few things to fix.

**Comment by szepeviktor** [comment]
> currently, phpstan reports 77 issues

Most of them take 20 seconds to fix.

**Comment by szepeviktor** [comment]
> > Please fix includes/class-plugin.php:207 first. That is a critical error.
> 
> you mean those two errors?

Yes.

**Comment by zaherg** [comment]
based on the [link](https://developer.wordpress.org/reference/functions/wp_add_dashboard_widget/), two are missing.

just pushed a fix

**Comment by szepeviktor** [comment]
> just pushed a fix

You are almost ready.

**Comment by zaherg** [comment]
> > just pushed a fix
> 
> You are almost ready.

Am I missing anything else?

**Comment by szepeviktor** [comment]
> Am I missing anything else?

No, the last commit is complete.

There is only half an hour work and the project will be error-free.

**Comment by zaherg** [comment]
I'll have to look at it later, its midnight here so its time for me to 😴

**Comment by szepeviktor** [comment]
WordPress core is not PHP 8.1 compatible.
The WordPress ecosystem is not PHP 8.0 compatible.
Actually I run every WP installation on PHP 7.4

Please consider running PHPStan on 8.0
I'm not sure!

**Comment by szepeviktor** [comment]
# 🚀

File: includes/class-plugin.php [hunk 1]
```diff
@@ -282,7 +282,7 @@ public function enqueue_admin_styles() {
         if ( ! in_array( $screen->id, $screens, true ) ) {
             return;
         }
-
+        /** @phpstan-ignore-next-line */
         wp_enqueue_style( 'redis-cache', WP_REDIS_PLUGIN_DIR . '/assets/css/admin.css', null, WP_REDIS_VERSION );
```
Comment (szepeviktor) [inline includes/class-plugin.php:286]: @zaherg Could an empty array in place of `null` fix this?
https://developer.wordpress.org/reference/functions/wp_enqueue_style/

Comment (zaherg) [reply includes/class-plugin.php]: just pushed the fix, I was reading the error from `wp_enqueue_script` while updating it 🫣 , sorry about that, WP is not my strongest feature

**Comment by zaherg** [comment]
I'll skip the CI from now on until I can confirm that I was able to fix the one I know how to fix

**Comment by szepeviktor** [comment]
> I'll skip the CI from now on

GHA won't run if you add `[skip ci]` to the commit message.

**Comment by zaherg** [comment]
> > I'll skip the CI from now on
> 
> GHA won't run if you add `[skip ci]` to the commit message.

This is what I was going to do with the future commits

**Comment by zaherg** [comment]
@szepeviktor do you like to have all the fix (the one I can) in one PR or we can now push this small fix, and then work on the others one by one, so I don't break the code ?

**Comment by szepeviktor** [comment]
Would you like me to send a PR to this branch resolving all problems reported?
It takes only couple of minutes.

**Comment by zaherg** [comment]
> Would you like me to send a PR to this branch resolving all problems reported? It takes only couple of minutes.

If you can, thats would be awesome

**Comment by szepeviktor** [comment]
> If you can, thats would be awesome

🏃🏻‍♂️

**Comment by szepeviktor** [comment]
Almost ready.

Please bring in Relay stubs.

**Comment by zaherg** [comment]
ok, one min

**Comment by szepeviktor** [comment]
Please review https://github.com/zaherg/redis-cache/pull/1

**Comment by szepeviktor** [comment]
> Found 23 errors

🎉

**Comment by szepeviktor** [comment]
Please spend 5 more minutes fixing some of those.
Then put all the remaining errors into a baseline file.

**Comment by zaherg** [comment]
I'll do it later today .. thanks @szepeviktor for all the help

File: includes/ui/tabs/overview.php [hunk 2]
```diff
@@ -7,6 +7,8 @@
 
 defined( '\\ABSPATH' ) || exit;
 
+/** @var \Rhubarb\RedisCache\Plugin $roc */
+
```
Comment (tillkruss) [inline includes/ui/tabs/overview.php:11]: Let's drop this space.

File: phpstan.neon.dist [hunk 3]
```diff
@@ -1,22 +1,33 @@
...
+    scanFiles:
+        - php-tools/relay.stub.php
+
     paths:
         - includes/
         - php-tools/object-cache.php
-    excludes_analyse:
+
+    excludePaths:
         - includes/object-cache.php
```
Comment (tillkruss) [inline phpstan.neon.dist:29]: Excluding `includes/object-cache.php` seems risky, @szepeviktor?

File: includes/object-cache.php [hunk 4]
```diff
@@ -2472,7 +2473,7 @@ public function switch_to_blog( $_blog_id ) {
             return false;
         }
 
-        $this->blog_prefix = $_blog_id;
+        $this->blog_prefix = (string) $_blog_id;
```
Comment (tillkruss) [inline includes/object-cache.php:2476]: I rather keep this a `(int)` across the codebase.

File: includes/diagnostics.php [hunk 5]
```diff
@@ -9,6 +9,8 @@
 
 global $wp_object_cache;
 
+/** @var \Rhubarb\RedisCache\Plugin $roc */
+
```
Comment (tillkruss) [inline includes/diagnostics.php:13]: Let's drop the extra line.

**Review by tillkruss — CHANGES_REQUESTED** [review]

File: phpstan.neon.dist (hunk 3 above)
Comment (zaherg) [reply phpstan.neon.dist:29]: Its the same as `php-tools/object-cache.php`, when we run `analyze` we create a copy of `includes/object-cache.php` do some magic (ref here https://github.com/rhubarbgroup/redis-cache/pull/371#issuecomment-1190724245) and run the analyze on it.

File: includes/ui/tabs/overview.php (hunk 2 above)
Comment (szepeviktor) [reply includes/ui/tabs/overview.php:11]: This "defines" `$roc`

I leave it 100% up to you.

File: includes/object-cache.php (hunk 4 above)
Comment (szepeviktor) [reply includes/object-cache.php:2476]: Okay! Much better!
Than please change type of `blog_prefix` property.

File: includes/diagnostics.php (hunk 5 above)
Comment (szepeviktor) [reply includes/diagnostics.php:13]: This "defines" $roc

I leave it 100% up to you.

File: includes/ui/tabs/overview.php (hunk 2 above)
Comment (szepeviktor) [reply includes/ui/tabs/overview.php:11]: When it touches the next line: there will be Confusion.

**Comment by szepeviktor** [comment]
"php-tools" was coined by me when PHPStan needed a higher PHP version than `redis-cache`

Now we keep only the renamed class file in that directory.
Could someone come up with a good name??

**Comment by zaherg** [comment]
> "php-tools" was coined by me when PHPStan needed a higher PHP version than `redis-cache`
> 
> Now we keep only the renamed class file in that directory. Could someone come up with a good name??

we can change it to `.includes` or `.phpstan`? as it is only used only within PHPStan

**Comment by szepeviktor** [comment]
People with more experience call this directory `build`

**Comment by zaherg** [comment]
I don't think `build` is the correct wording here, but for me any wording works, at the end its not part of the core code, but I'll leave it up to you and Till to decide what to call it.

**Comment by tillkruss** [comment]
@zaherg: Let me know when you're done and I'll take a look at the remaining issues.

**Comment by zaherg** [comment]
@tillkruss i don't think I have anything else to add to it.

**Pull request merged**